src/misc/mixing.c

```diff
@@ -576,3 +576,322 @@ igraph_error_t igraph_joint_degree_matrix(const igraph_t* graph,
+    igraph_vector_int_t *deg_from, *deg_to, deg_out, deg_in, deg_all;
+
+    /* Make sure directionality parameters are consistent for undirected graphs. */
+    if (! igraph_is_directed(graph)) {
+        from_mode = to_mode = IGRAPH_ALL;
+        directed_neighbors = false;
+    }
+
+    igraph_bool_t have_out = from_mode == IGRAPH_OUT || to_mode == IGRAPH_OUT;
+    igraph_bool_t have_in  = from_mode == IGRAPH_IN  || to_mode == IGRAPH_IN;
+    igraph_bool_t have_all = from_mode == IGRAPH_ALL || to_mode == IGRAPH_ALL;
+
+    if (have_out) {
+        IGRAPH_VECTOR_INT_INIT_FINALLY(&deg_out, no_of_nodes);
+        IGRAPH_CHECK(igraph_degree(graph, &deg_out, igraph_vss_all(), IGRAPH_OUT, /* loops */ true));
+    }
+
+    if (have_in) {
+        IGRAPH_VECTOR_INT_INIT_FINALLY(&deg_in, no_of_nodes);
+        IGRAPH_CHECK(igraph_degree(graph, &deg_in, igraph_vss_all(), IGRAPH_IN, /* loops */ true));
+    }
+
+    if (have_all) {
+        IGRAPH_VECTOR_INT_INIT_FINALLY(&deg_all, no_of_nodes);
+        IGRAPH_CHECK(igraph_degree(graph, &deg_all, igraph_vss_all(), IGRAPH_ALL, /* loops */ true));
+    }
+
+    switch (from_mode) {
+    case IGRAPH_OUT: deg_from = &deg_out; break;
+    case IGRAPH_IN:  deg_from = &deg_in;  break;
+    case IGRAPH_ALL: deg_from = &deg_all; break;
+    default:
+        IGRAPH_ERROR("Invalid 'from' degree mode.", IGRAPH_EINVAL);
+    }
+
+    switch (to_mode) {
+    case IGRAPH_OUT: deg_to = &deg_out; break;
+    case IGRAPH_IN:  deg_to = &deg_in;  break;
+    case IGRAPH_ALL: deg_to = &deg_all; break;
+    default:
+        IGRAPH_ERROR("Invalid 'to' degree mode.", IGRAPH_EINVAL);
+    }
+
+    IGRAPH_CHECK(mixing_matrix(graph, p,
+                               deg_from, deg_to, max_from_degree, max_to_degree,
+                               directed_neighbors, normalized,
+                               weights, /*check_types=*/ false));
+
+    if (have_all) {
+        igraph_vector_int_destroy(&deg_all);
+        IGRAPH_FINALLY_CLEAN(1);
+    }
+
+    if (have_in) {
+        igraph_vector_int_destroy(&deg_in);
+        IGRAPH_FINALLY_CLEAN(1);
+    }
+
+    if (have_out) {
+        igraph_vector_int_destroy(&deg_out);
+        IGRAPH_FINALLY_CLEAN(1);
+    }
+
+    return IGRAPH_SUCCESS;
+}
+
+
+/**
+ * \function igraph_mixing_matrix
+ * \brief Mixing matrix for vertex categories.
+ *
+ * Computes the mixing matrix M_ij, i.e. the joint distribution of vertex categories.
+ * Categories are represented by non-negative integer indices, passed in
+ * \p from_types and \p to_types. The row and column counts of \p m will be one
+ * larger than the largest source and target type, respectively. Re-index type
+ * vectors using \ref igraph_reindex_membership() if they are not contiguous integers,
+ * to avoid producing a very large matrix.
+ *
+ * </para><para>
+ * M_ij is proportional to the probability that a randomly chosen ordered vertex pair
+ * have types \c i and \c j.
```

Comment:
```suggestion
 * has types \c i and \c j.
```

Reply:
This was intentional, but now it is rephrased.